lib/propolis/src/block/mod.rs

```diff
@@ -650,3 +650,260 @@ impl DriverInner {
+    }
+
+    fn start(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.run();
+        });
+    }
+    fn pause(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.hold();
+        });
+    }
+    fn resume(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.run();
+        });
+    }
+    fn halt(&self) {
+        self.with_ctrl(|ctrl| {
+            let _ = ctrl.exit();
+        });
+    }
+
+    fn worker(self: &Arc<Self>) -> WorkerHdl {
+        WorkerHdl::new(self.clone())
+    }
+
+    /// Attempt to grab a request from the block device (if one's available)
+    async fn next_req(&self) -> Request {
+        let bdev = self.bdev.get().expect("attached block device");
+        loop {
+            if let Some(req) = bdev.next() {
+                return req;
+            }
+
+            // Don't busy-loop on the device but just wait for it to wake us
+            // when the next request is available
+            self.wake.notified().await;
+        }
+    }
+
+    fn complete_req(&self, req: Request, res: Result) {
+        let bdev = self.bdev.get().expect("attached bdev");
+        req.complete(res, bdev.as_ref());
+        match self.active_req_count.fetch_sub(1, Ordering::SeqCst) {
+            0 => panic!("request completion count mismatch"),
+            1 => self.wake.notify_one(),
+            _ => {}
+        }
+    }
+
+    /// Scheduling task body: feed worker threads with requests from block device.
+    async fn schedule_requests(&self, task: &mut TaskHdl) {
+        loop {
+            let avail = tokio::select! {
+                event = task.get_event() => {
+                    match event {
+                        Event::Hold => {
+                            // Wait for all pending requests to complete
+                            let areq = &self.active_req_count;
+                            while areq.load(Ordering::Acquire) != 0 {
+                                self.wake.notified().await;
+                            }
+
+                            // ... then enter held state
+                            task.wait_held().await;
+                            continue;
+                        }
+                        Event::Exit => {
+                            // Close the semaphore to indicate to the workers
+                            // that they should bail out too
+                            self.worker_activate.close();
+                            return;
+                        }
+                    }
+                },
+                avail = self.worker_idle.acquire() => {
+                    // We found an idle thread!
+                    avail.unwrap()
+                }
+            };
+
+            tokio::select! {
+                _event = task.get_event() => {
+                    // Take another lap if an event arrives while we are waiting
+                    // for a request to schedule to the worker
```

Comment:
Worth adding a note here that we're not handling the event explicitly here is fine since `get_event` is essentially level-triggered?

Reply:
Updated the comment with more detail on that subject